### PR TITLE
[3.10.9] Links from numeric tags cause error 404 #37814

### PR DESCRIPTION
**Pull request by conseilgouz**
Pull Request for Issue #37814 .

### Testing Instructions

Create a numeric tag, a non-numeric tag starting with a number and a non-numeric tag starting with a non-numeric character in one article

### Actual result BEFORE applying this Pull Request

Non numeric tags articles list displayed when selecting a non-numeric tag
error 404 when clicking on numeric tag or a non-numeric tag starting with a number.

Same results in popular tags module

### Expected result AFTER applying this Pull Request

Non numeric tags articles list displayed when selecting a non-numeric tag
when clicking on numeric tag, articles list is displayed
when clicking on non-numeric tag starting with a number, articles list is displayed

Popular tags module works fine in all cases

### CODE REVIEW

**Comment by conseilgouz** [comment]
same syntax than com_tags/views/tag/view.feed.php

File: components/com_tags/models/tag.php [hunk 1]
```diff
@@ -178,7 +178,7 @@ protected function populateState($ordering = 'c.core_title', $direction = 'ASC')
 		$this->setState('params', $params);
 
 		// Load state from the request.
-		$ids = (array) $app->input->get('id', array());
+		$ids = (array) $app->input->get('id', array(),'int');
```
Comment (zero-24) [inline components/com_tags/models/tag.php]:
```suggestion
		$ids = (array) $app->input->get('id', array(), 'int');
```

Comment (richard67) [inline components/com_tags/models/tag.php]:
```suggestion
		$ids = (array) $app->input->get('id', array(), 'int');
```
Fix code style error.

**Comment by zero-24** [comment]
@conseilgouz please apply the change as suggested by @richard67 and me to fix the drone issue: https://ci.joomla.org/joomla/joomla-cms/54145/1/3

**Comment by richard67** [comment]
@conseilgouz It's not enough to mark suggested changes as resolved for applying the changes. It needs to commit a suggestion to apply it, or if you don't want to apply it then tell the reason in an answering comment. But not just mark the suggestion as resolved without doing anything.

**Comment by toivo** [comment]
I have tested this item :white_check_mark: successfully on 5b75a954453e0ea9eb64a09908e04e840f89587c<br /><br />Tested successfully in Joomla 3.10.10-dev of 17 May using PHP 8.0.15<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37815">issues.joomla.org/tracker/joomla-cms/37815</a>.</sub>

**Comment by richard67** [comment]
@toivo Have you also tested that links from a non-numerical tag still work?

@conseilgouz I think your testing instructions should be completed by such a step to test if non-numerical tags are not broken. In general it is not sufficient to test only if a PR solves an issue, it needs to test also if the function is not broken for other use cases.

**Comment by eeshaanSA** [comment]
i would like to work on this issue. This would potentially be my first contribution to the Joomla! CMS. Would need some help to pull it off.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37815">issues.joomla.org/tracker/joomla-cms/37815</a>.</sub>

**Comment by eeshaanSA** [comment]
please tell me how can i test this issue and help solve it.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37815">issues.joomla.org/tracker/joomla-cms/37815</a>.</sub>

**Comment by richard67** [comment]
> i would like to work on this issue. This would potentially be my first contribution to the Joomla! CMS. Would need some help to pull it off.

@eeshaanSA This here is not an issue, it is already a pull request for solving the issue. So @conseilgouz has already worked on it, and this here is the result. You can help with testing it. But it would not make much sense to have 2 people working on the same issue for which there is already a solution (= this pull request here).

**Comment by toivo** [comment]
@richard67 yes, non-numeric tags also work as links.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37815">issues.joomla.org/tracker/joomla-cms/37815</a>.</sub>

File: components/com_tags/models/tag.php (hunk 1 above)
Comment (zero-24) [inline components/com_tags/models/tag.php]:
```suggestion
		$ids = (array) $app->input->get('id', array(), 'int');
```

**Comment by toivo** [comment]
I have tested this item :white_check_mark: successfully on 23d232148a5515ca5d8c4cc9138bb0592d03bbdb<br /><br />Tested successfully in Joomla 3.10.10-dev of 18 May using PHP 8.0.15 in Wampserver 3.2.8.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37815">issues.joomla.org/tracker/joomla-cms/37815</a>.</sub>

**Comment by conseilgouz** [comment]
@eeshaanSA : you could test this pr as one more human test is required.

See https://docs.joomla.org/Testing_Joomla!_patches/en

**Comment by Magnytu2** [comment]
Hello, I think I have a similar problem on Joomla 4.1.3. php 7.4
Some tags make a return to the home page and others give a 404 error.
https://planetarium-epinal.com/etablissement-scolaire/les-ateliers-scientifiques

**Comment by conseilgouz** [comment]
> Hello, I think I have a similar problem on Joomla 4.1.3. php 7.4 Some tags make a return to the home page and others give a 404 error. https://planetarium-epinal.com/etablissement-scolaire/les-ateliers-scientifiques

It's the same issue in Joomla 4.

In your case, it's a tag starting with a number.

I just tested it with Joomla 3.10.9 and everything is fine. 

Same fix should be done in J.4.1 (it's in line 189)

**Comment by conseilgouz** [comment]
I just created a 4.1 PR : #37856

**Comment by zero-24** [comment]
Usually there is no need to do a PR against 4.1 about the same issue with a downstream branch once we get the seccond test here the patch will be merged and will be merged up to 4.1 and 4.2

**Comment by Magnytu2** [comment]
Ok good. I am relieved that a solution can be quickly deployed. But why do the other tags refer to the home page instead of displaying the articles common to the tag?

**Comment by zero-24** [comment]
@Magnytu2 can you confirm that the issue has been fixed by the change and mark this as a successful test? If so it can be merged and fixed with the next releases ;)

**Comment by fontanil** [comment]
I tested this patch on a 3.10.9 version: it works fine for me.

**Comment by richard67** [comment]
@fontanil Could you mark your test result in the issue tracker? Just go to https://issues.joomla.org/tracker/joomla-cms/37815 , then use the blue “Test this” button at the top left corner, select your test result and then submit. Thanks in advance.

**Comment by fontanil** [comment]
I have tested this item :white_check_mark: successfully on 23d232148a5515ca5d8c4cc9138bb0592d03bbdb<br /><br />Tested successfully on Joomla! 3.10.9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37815">issues.joomla.org/tracker/joomla-cms/37815</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37815">issues.joomla.org/tracker/joomla-cms/37815</a>.</sub>

**Comment by joomdonation** [comment]
By just reading the code, the change here could be backward-incompatible. Before this change, we allow displaying multiple tags by passing tag IDs comma separated like this http://localhost/joomla4/index.php?option=com_tags&view=tag&id=2,3, but with the change introduced by this PR, this behavior is lost

What is actual happening here is that in our routing code, we return tag id in this form **ID:ALIAS** (see https://github.com/joomla/joomla-cms/blob/3.10-dev/components/com_tags/router.php#L183 - I don't know why we have code like that) 

Then on this code https://github.com/joomla/joomla-cms/blob/4.1-dev/components/com_tags/src/Model/TagModel.php#L189 , we use `cmd` filter, the **:** character is removed and later, when it is converted to `int`, it return wrong value for numeric tags (while it is still OK for string tags)

Without knowing how links to tags are used on all places in Joomla, I think it would be more safe to use `string` filter instead of `int` here. It solves the issue, and won't cause by unexpected (backward-incompatible) behavior. Ping @zero-24

**Comment by richard67** [comment]
Back to pending due to reasons stated above.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37815">issues.joomla.org/tracker/joomla-cms/37815</a>.</sub>

**Comment by joomdonation** [comment]
I have tested this item :white_check_mark: successfully on 79bce845dfa59f2b4842ca8c32cacbc906f564a2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37815">issues.joomla.org/tracker/joomla-cms/37815</a>.</sub>

**Comment by joomdonation** [comment]
@toivo or @fontanil Could you please re-test?

**Comment by fontanil** [comment]
I have tested this item :white_check_mark: successfully on 79bce845dfa59f2b4842ca8c32cacbc906f564a2<br /><br />Tested successfully. Thanks!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37815">issues.joomla.org/tracker/joomla-cms/37815</a>.</sub>

**Comment by Magnytu2** [comment]
The patch does not seem to work.
Capture d’écran 2022-05-23 à 09 45 01

**Comment by richard67** [comment]
@Magnytu2 You can't apply a PR for 3.10 on a 4.x if the file which is changed is at a different place in 3.10 and 4.x.

**Comment by richard67** [comment]
P.S.: And in general you should not try to apply a PR for 3.10 on a 4.x and vice versa.

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37815">issues.joomla.org/tracker/joomla-cms/37815</a>.</sub>

**Comment by Magnytu2** [comment]
I do not understand. It's apparently the same PR, but I can't test it? And so fix my website?

**Comment by Magnytu2** [comment]
And that does not answer my question why my tags return to the homepage or in error 404...? Is it the same problem?

**Comment by richard67** [comment]
> I do not understand. It's apparently the same PR, but I can't test it? And so fix my website?

@Magnytu2 Your screenshot shows that you are trying to apply this PR which is made for 3.10 on a Joomla 4 installation. This cannot work, especially because in Joomla 3 the changed file is https://github.com/joomla/joomla-cms/blob/3.10-dev/components/com_tags/models/tag.php and in Joomla 4 it is https://github.com/joomla/joomla-cms/blob/4.1-dev/components/com_tags/src/Model/TagModel.php , so it is obviously at a different place, so the patchtester can not apply the change on a Joomla 4.

It should be clear by my previous comment. I have no idea what was not understandable.

**Comment by richard67** [comment]
@Magnytu2 P.S.: The PR which you have tested before was not this one but the one for 4.1-dev , PR #37856 . If you want to fix your site you can manually change the file modified by that PR in the way this PR here changes it for J3.

**Comment by Magnytu2** [comment]
Sorry, what is not easy to understand is that we use a single PR for two versions of Joomla and therefore we can only test on a single version.

**Comment by richard67** [comment]
> Sorry, what is not easy to understand is that we use a single PR for two versions of Joomla and therefore we can only test on a single version.

@Magnytu2 It seems you are the only one who has problems to understand that. This PR here will be merged into 3.10-dev and then the change is merged up into 4.1-dev, where it goes into the right file at the right place, like we did it all the time.

**Comment by Magnytu2** [comment]
Yes excuse me. I don't speak English, and I don't have the skills to modify a Php file. But my site relies on the use of tags and it is very important. So two more questions please. Can I switch my joomla 4.1 to DEV version to get the patches and if so from when? Because I also stay with a use of my tags which no longer makes a list of articles but which sends me back to the home page. Thank you

**Comment by richard67** [comment]
> Yes excuse me. I don't speak English, and I don't have the skills to modify a Php file. But my site relies on the use of tags and it is very important. So two more questions please. Can I switch my joomla 4.1 to DEV version to get the patches and if so from when? Because I also stay with a use of my tags which no longer makes a list of articles but which sends me back to the home page. Thank you

@Magnytu2 I would not recommend to switch a live site to development versions. But you can do that on a copy of your life site either on a local webserver or on a subdomain of your domain. There you could use the nightly builds.

For now, to solve your issue, I offer you to download this zip file from my domain: https://test5.richard-fath.de/TagModel.zip .

It contains the modified "TagModel.php" file with the fix from this PR here.

Unpack the zip and then upload the TagModel.php file into folder "components/com_tags/src/Model" of your website.

**Comment by Magnytu2** [comment]
Oh awesome! A HUGE thank you. I no longer get 404 errors on tags. I always have the problem of redirecting to the homepage that I don't understand when I click on a tag. But I'm looking.

**Comment by zero-24** [comment]
Merging here to please @richard67 . Thanks @conseilgouz

**Pull request merged**